BASIC NUMPY:
---------------------
(This is a recreation following this: https://sebastianraschka.com/blog/2020/numpy-intro.html)

In [2]:
import numpy as np

In [4]:
print(np.arange(6))

[0 1 2 3 4 5]


In [3]:
def python_dot(x, w):
    z = 0.
    for i in range(len(x)):
        z+= x[i]*w[i]
    return z

In [6]:
array1 = [1., 2., 5.]
array2 = [3., 1., 2.]

In [7]:
print(python_dot(array1, array2))

15.0


Now, let us time a large array:

In [8]:
large1 = list(range(1000))
large2 = list(range(1000))

In [9]:
%timeit python_dot(large1, large2)

59.5 µs ± 1.6 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


* Now, if we use np arrays:*

In [4]:
nplarge1 = np.arange(1000)
nplarge2 = np.arange(1000)

In [5]:
%timeit nplarge1.dot(nplarge2)

1.36 µs ± 45.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


We observe 59.5-->1.36. nparrays are like C arrays, contiguous and homogenous, Helps in cache locality

________________________________________________________________________________________________________

Now, lets look at multidimensional arrays:

In [6]:
pyarray = [1., 2., 3.]

In [7]:
numpyarray = np.array(pyarray)

In [8]:
print(type(numpyarray))

<class 'numpy.ndarray'>


In [9]:
print(numpyarray)

[1. 2. 3.]


In [15]:
pyListOList = [ [1,2,3],
                [4,5,6]]

In [16]:
mat = np.array(pyListOList)

In [17]:
mat = np.array(pyListOList, dtype=int)

In [18]:
type(mat)

numpy.ndarray

In [19]:
print(mat.dtype)

int64


In [21]:
mat_int32 = mat.astype(np.int32)
print(mat_int32.dtype)
type(mat_int32)

int32


numpy.ndarray

Like above, we can convert dtype

In Python, all variables are like references in C++. So, if you want just the value..

In [24]:
print(mat)
mat2 = mat
mat2[0][0] = 5
print(mat)
mat_copy = mat.copy()
mat_copy[0][0] = 4
print(mat)

[[3 2 3]
 [4 5 6]]
[[5 2 3]
 [4 5 6]]
[[5 2 3]
 [4 5 6]]


We have to use ndarry.copy()

nparrays have ndim attribute:

In [25]:
mat.ndim

2

The Shape Attribute: returns nxm as (n, m) {returns a tuple(in case of ndim, so tuple)}

In [26]:
print(mat.shape)

(2, 3)


In [27]:
print(len(mat.shape))

2


In [30]:
assert(len(mat.shape)==mat.ndim)

So, mat.ndim = the length of the shape tuple
_________________________________________________________________________________


NUMPY ARRAY CONSTRUCTION:
-------------------------
Lets look at some array construction routines